<a href="https://colab.research.google.com/github/JialinKang/hello-world/blob/master/Question4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Fit a multilayer neural network model on the SHHS data to predict whether rdi4p > 7 using the variables from the previous HW assigment. Compare the test set accuracy with the penalized model from your previous HW.**

In [0]:
from sklearn.model_selection import train_test_split
import keras
import pandas as pd
import numpy as np
ssh_data = pd.read_table('https://raw.githubusercontent.com/JialinKang/hello-world/master/shhs1.txt')

In [0]:
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers.core import Dense, Dropout, Activation
from keras.utils import np_utils
from keras import layers

In [6]:
j = 0
ssh_data['predict'] = np.zeros((len(ssh_data,)))
for i in ssh_data['rdi4p']:
  if i <=7 :
    ssh_data['predict'][j] = 0
  else:
    ssh_data['predict'][j] = 1
  j = j+1

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  import sys


In [0]:
ssh_data.dropna(axis=0)
for i in list(ssh_data):
    ssh_data = ssh_data[~np.isnan(ssh_data[i])]

In [0]:
label = ssh_data['predict']
ssh_data=ssh_data.drop(['pptid','rdi4p','predict'],axis=1)
x_train, x_test, y_train, y_test = train_test_split(ssh_data, label, test_size=0.2, random_state=2)

In [11]:
model = Sequential() # creat model
model.add(Dense(64,activation='relu',input_shape=(28,)))

model.add(Dropout(0.2))
model.add(Dense(64,activation='relu'))
model.add(Dropout(0.2))
model.add(Dense(1,activation='sigmoid'))
model.compile(optimizer='adam',loss='binary_crossentropy',metrics=['accuracy'])

model.fit(x_train,y_train,batch_size=300,epochs=50,validation_data=(x_test,y_test)) #train the model and test data for validation
score = model.evaluate(x_test,y_test)
print("loss:",score[0])
print("accu:",score[1])#print accuracy

Train on 2317 samples, validate on 580 samples
Epoch 1/50
2317/2317 [==============================] - 0s 154us/step - loss: 20.1874 - accuracy: 0.4644 - val_loss: 11.6203 - val_accuracy: 0.6552
Epoch 2/50
2317/2317 [==============================] - 0s 13us/step - loss: 14.2358 - accuracy: 0.5896 - val_loss: 5.9610 - val_accuracy: 0.6362
Epoch 3/50
2317/2317 [==============================] - 0s 11us/step - loss: 10.0882 - accuracy: 0.5416 - val_loss: 2.9980 - val_accuracy: 0.5052
Epoch 4/50
2317/2317 [==============================] - 0s 11us/step - loss: 9.4923 - accuracy: 0.5166 - val_loss: 3.7024 - val_accuracy: 0.6690
Epoch 5/50
2317/2317 [==============================] - 0s 11us/step - loss: 7.7183 - accuracy: 0.5697 - val_loss: 2.3273 - val_accuracy: 0.6379
Epoch 6/50
2317/2317 [==============================] - 0s 11us/step - loss: 6.9782 - accuracy: 0.5343 - val_loss: 1.7232 - val_accuracy: 0.5828
Epoch 7/50
2317/2317 [==============================] - 0s 12us/step - loss: 5

In [13]:
from sklearn.linear_model import LogisticRegression
import numpy as np

# Logsitic regression with penalty
logisticRegr = LogisticRegression(penalty='l1', solver='saga')
logisticRegr.fit(x_train, y_train)
l1score = logisticRegr.score(x_test, y_test)
print('Accuracy for logistic regression model with lasso penalty: ' + str(l1score))
logisticRegr = LogisticRegression(penalty='l2', solver='saga')
logisticRegr.fit(x_train, y_train)
l2score = logisticRegr.score(x_test, y_test)
print('Accuracy for logistic regression model with ridge penalty: ' + str(l2score))
logisticRegr = LogisticRegression(penalty='elasticnet', solver='saga', l1_ratio=0.2)
logisticRegr.fit(x_train, y_train)
enscore = logisticRegr.score(x_test, y_test)
print('Accuracy for logistic regression model with elasticnet penalty: ' + str(enscore))

/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


Accuracy for logistic regression model with lasso penalty: 0.6810344827586207
Accuracy for logistic regression model with ridge penalty: 0.6810344827586207
Accuracy for logistic regression model with elasticnet penalty: 0.6810344827586207


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_sag.py:330: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


From above result, we can know:

NN model accuacy: 0.6603

Accuracy for logistic regression model with lasso penalty: 0.6810344827586207

Accuracy for logistic regression model with ridge penalty: 0.6810344827586207

Accuracy for logistic regression model with elasticnet penalty: 0.6810344827586207

So in this case the logistic regression model is better than NN model